internal/pkg/agent/cmd/container.go

```diff
@@ -988,3 +991,129 @@ func isContainer(detail component.PlatformDetail) component.PlatformDetail {
 	detail.OS = component.Container
 	return detail
 }
+
+// agentInfo implements the AgentInfo interface, and it used in shouldFleetEnroll.
+type agentInfo struct {
+	id string
+}
+
+func (a *agentInfo) AgentID() string {
+	return a.id
+}
+
+// shouldFleetEnroll returns true if the elastic-agent should enroll to fleet.
+func shouldFleetEnroll(setupCfg setupConfig) (bool, error) {
+	if !setupCfg.Fleet.Enroll {
+		// fleet enrollment is not enabled
+		return false, nil
+	}
+
+	if setupCfg.Fleet.Force {
+		// fleet enrollment is enforced
+		return true, nil
+	}
+
+	agentCfgFilePath := paths.AgentConfigFile()
+	_, err := os.Stat(agentCfgFilePath)
+	if os.IsNotExist(err) {
+		// config file does not exist so elastic-agent should enroll
+		return true, nil
+	}
+
+	// read the stored elastic-agent state to extract the previous fleet url and enrollment token
+	ctx := context.Background()
+	store, err := storage.NewEncryptedDiskStore(ctx, agentCfgFilePath)
+	if err != nil {
+		return false, fmt.Errorf("failed to instantiate encrypted disk store: %w", err)
+	}
+
+	reader, err := store.Load()
+	if err != nil {
+		return false, fmt.Errorf("failed to load from disk store: %w", err)
+	}
+
+	cfg, err := config.NewConfigFrom(reader)
+	if err != nil {
+		return false, fmt.Errorf("failed to read from disk store: %w", err)
+	}
+
+	storedConfig, err := configuration.NewFromConfig(cfg)
+	if err != nil {
+		return false, fmt.Errorf("failed to read from disk store: %w", err)
+	}
+
+	storedFleetHosts := storedConfig.Fleet.Client.GetHosts()
+	if len(storedFleetHosts) == 0 || !slices.Contains(storedFleetHosts, setupCfg.Fleet.URL) {
+		// fleet url changed thus elastic-agent should enroll
+		return true, nil
```

Comment:
What if they are just updating the endpoint at which Fleet Server is running but its the same Fleet Server? I would expect just checking the token hash is enough, because if it is a different Fleet Server then token would 100% be different.

Reply:
are you talking of changing solely the DNS records?! because if they actually change the endpoint IP then using the fleet url from the state at run the agent won't be able to connect to it right?!

Reply:
Not if they setup a redirect and then also updated the policy with a new URL. Then it will start connect be redirected, connect get new policy revision and then update the local state. Then they update the container manifest and then remove the redirect.

Reply:
If I follow the flow correctly this is something that will happen through the Kibana-Fleet UI where the agent will persist the changes to it's state without restarting. So the debate here is what happens when the user restarts the pod as the manifest will still point to the old url?

Reply:
The flow is both happen. They update the URL in Kibana, then they update the manifest for the Elastic Agent. They would not want it to re-enroll its already enrolled.

Reply:
> They update the URL in Kibana, then they update the manifest for the Elastic Agent.

if they add the same URL from Kibana and then in the manifest, then they won't re-enroll. The re-enrollment will happen only if the URL defined in the manifest doesn't exist in the hosts saved in the state, right?

Reply:
That is correct. I was mis-reading this. I was thinking that it was caching the enrollment URL, but actually this is coming from the policy hosts. This makes more sense, sorry for the noise.

Reply:
I updated the comment on this part in an attempt to make that more visible. Feel free to unresolve if you spot anything else 🙂